### PR TITLE
Added the verification for reservations overlapping

### PR DESCRIPTION
**Pull request by Faka-GLB**
Created a ReservationVerifier class that checks the startDate and endDate Calendars of the new reservation doesn't overlap an existing one in the ReservationsDatabase. 
There are 3 cases of overlapping:
- Start date overlap or start and end overlap: the new reservation start date is within the saved one, this also checks start and end case.
- End date overlap: the new reservation end date overlaps into the saved one 
-Saved reservation overlaps the new one: a new reservation time frame contains the saved one
 
If any of this cases are true then a toast tells the user that an overlap occurs letting the user switch the dates and times in de DateTimePicker. Otherwise the reservation doesn't overlap and is stored in the database.

project tree

### CODE REVIEW

File: app/src/main/java/com/onboarding/parkingsystemjava/utils/ReservationVerifier.java [hunk 1]
```diff
@@ -0,0 +1,58 @@
...
+    }
+
+    private boolean isOverlapping(Reservation reservation, Reservation anotherReservation) {
+        Calendar reservationStartDate = reservation.getStartDate();
+        Calendar reservationEndDate = reservation.getEndDate();
+        Calendar anotherReservationStartDate = anotherReservation.getStartDate();
+        Calendar anotherReservationEndDate = anotherReservation.getEndDate();
+
+        //start or full overlap
+        if (reservationStartDate.before(anotherReservationStartDate) &&
```
Comment (BQuinteros) [inline app/src/main/java/com/onboarding/parkingsystemjava/utils/ReservationVerifier.java:47]: pls use guard clauses

File: app/src/main/java/com/onboarding/parkingsystemjava/mvp/model/ReserveModel.java [hunk 2]
```diff
@@ -49,8 +52,12 @@ else if (reservation.getParkingLot() == ConstantUtils.INT_MINUS_ONE)
             return MISSING_LOT;
         else if (reservation.getUserPassword().isEmpty())
             return MISSING_PASSWORD;
-        else
+        else if (verifier.isReservationOK(this.reservation)) {
```
Comment (BQuinteros) [inline app/src/main/java/com/onboarding/parkingsystemjava/mvp/model/ReserveModel.java]: pls use guard clauses

Comment (BQuinteros) [reply app/src/main/java/com/onboarding/parkingsystemjava/mvp/model/ReserveModel.java]: https://refactoring.com/catalog/replaceNestedConditionalWithGuardClauses.html

File: app/src/main/java/com/onboarding/parkingsystemjava/mvp/model/ReserveModel.java [hunk 3]
```diff
@@ -23,7 +25,7 @@ public ReserveModel(ParkingDatabase database) {
 
     @Override
     public void addReservation(Reservation reservation) {
-        database.addReservations(reservation);
+        database.addReservation(reservation.getParkingLot(), reservation);
```
Comment (lucianoeli) [inline app/src/main/java/com/onboarding/parkingsystemjava/mvp/model/ReserveModel.java]: why this change?
lwas better before, the lotNumb is inside de reservation. use only 1 param

File: app/src/main/java/com/onboarding/parkingsystemjava/mvp/model/ReserveModel.java [hunk 4]
```diff
@@ -41,6 +43,7 @@ public void setEndDate(Calendar endDate) {
     }
 
     public ReserveComprobation reserveFieldsCheck(Reservation reservation) {
+        ReservationVerifier verifier = new ReservationVerifier(ParkingDatabase.getInstance());
```
Comment (lucianoeli) [inline app/src/main/java/com/onboarding/parkingsystemjava/mvp/model/ReserveModel.java]: you shoud instantiate this on the presenter, and passit as a parameter for the model. 
this way you will can mockit when testing
check dependency inversion

File: app/src/main/java/com/onboarding/parkingsystemjava/mvp/model/ReserveModel.java (hunk 2 above)
Comment (lucianoeli) [reply app/src/main/java/com/onboarding/parkingsystemjava/mvp/model/ReserveModel.java]: i think this checker could also check all previous checks, and return a code/constant/enum, with the type of response

File: app/src/main/java/com/onboarding/parkingsystemjava/utils/ReservationVerifier.java [hunk 5]
```diff
@@ -0,0 +1,58 @@
...
+                //stored reservation full overlaps new reservation
+                return reservationStartDate.after(anotherReservationStartDate) &&
+                        reservationEndDate.before(anotherReservationEndDate);
+            }
+        }
+    }
+
+    private boolean checkArrayList(ArrayList<Reservation> arrayList, Reservation reservation) {
+        Reservation storedReservation;
+        for (int i = ConstantUtils.INT_ZERO; i < arrayList.size(); i++) {
```
Comment (lucianoeli) [inline app/src/main/java/com/onboarding/parkingsystemjava/utils/ReservationVerifier.java:64]: can use .foreach

File: app/src/main/java/com/onboarding/parkingsystemjava/entity/Reservation.java [hunk 6]
```diff
@@ -10,7 +10,7 @@
     private String userPassword;
 
     public Reservation() {
-        this.parkingLot = - ConstantUtils.INT_ONE;
+        this.parkingLot = -ConstantUtils.INT_ONE;
```
Comment (mauroLonghi) [inline app/src/main/java/com/onboarding/parkingsystemjava/entity/Reservation.java]: plase avoid  "-" , use a real constant

Comment (lucianoeli) [reply app/src/main/java/com/onboarding/parkingsystemjava/entity/Reservation.java]: you already have `INT_MINUS_ONE `

File: app/src/main/java/com/onboarding/parkingsystemjava/mvp/model/ReserveModel.java (hunk 3 above)
Comment (Faka-GLB) [reply app/src/main/java/com/onboarding/parkingsystemjava/mvp/model/ReserveModel.java]: database is now a HashMap<int,Reservation>, so the parkingLot is the key, it still in the reservation, previously database was an ArrayList<Reservation>

File: app/src/main/java/com/onboarding/parkingsystemjava/utils/ReservationVerifier.java (hunk 1 above)
Comment (Faka-GLB) [reply app/src/main/java/com/onboarding/parkingsystemjava/utils/ReservationVerifier.java:47]: done

File: app/src/main/java/com/onboarding/parkingsystemjava/mvp/model/ReserveModel.java (hunk 2 above)
Comment (Faka-GLB) [reply app/src/main/java/com/onboarding/parkingsystemjava/mvp/model/ReserveModel.java]: done, replaced if/else nodes with guard clauses and moved fields check to ReservationVerifier

File: app/src/main/java/com/onboarding/parkingsystemjava/mvp/model/ReserveModel.java (hunk 4 above)
Comment (Faka-GLB) [reply app/src/main/java/com/onboarding/parkingsystemjava/mvp/model/ReserveModel.java]: done

File: app/src/main/java/com/onboarding/parkingsystemjava/entity/Reservation.java (hunk 6 above)
Comment (Faka-GLB) [reply app/src/main/java/com/onboarding/parkingsystemjava/entity/Reservation.java]: I've forgot about that, fixed

File: app/src/main/java/com/onboarding/parkingsystemjava/mvp/model/ReserveModel.java (hunk 3 above)
Comment (Faka-GLB) [reply app/src/main/java/com/onboarding/parkingsystemjava/mvp/model/ReserveModel.java]: done

File: app/src/main/java/com/onboarding/parkingsystemjava/utils/ReservationVerifier.java (hunk 5 above)
Comment (Faka-GLB) [reply app/src/main/java/com/onboarding/parkingsystemjava/utils/ReservationVerifier.java:64]: requieres apimin 24, and the project has apimin 23

File: app/src/main/java/com/onboarding/parkingsystemjava/utils/ReservationVerifier.java [hunk 7]
```diff
@@ -30,23 +45,20 @@ private boolean isOverlapping(Reservation reservation, Reservation anotherReserv
...
+        else {
+            //stored reservation full overlaps new reservation
+            return reservationStartDate.after(anotherReservationStartDate) &&
+                    reservationEndDate.before(anotherReservationEndDate);
         }
     }
 
     private boolean checkArrayList(ArrayList<Reservation> arrayList, Reservation reservation) {
         Reservation storedReservation;
+        //arrayList.forEach();
```
Comment (lucianoeli) [inline app/src/main/java/com/onboarding/parkingsystemjava/utils/ReservationVerifier.java]: delete comments please

Comment (Faka-GLB) [reply app/src/main/java/com/onboarding/parkingsystemjava/utils/ReservationVerifier.java]: done

File: app/src/main/java/com/onboarding/parkingsystemjava/activity/ReserveActivity.java [hunk 8]
```diff
@@ -21,7 +22,7 @@ protected void onCreate(@Nullable Bundle savedInstanceState) {
         super.onCreate(savedInstanceState);
         this.binding = ActivityNewReservationBinding.inflate(getLayoutInflater());
         setContentView(binding.getRoot());
-        this.presenter = new ReservePresenter(new ReserveView(this, binding), new ReserveModel(ParkingDatabase.getInstance()));
+        this.presenter = new ReservePresenter(new ReserveView(this, binding), new ReserveModel(ParkingDatabase.getInstance(), new ReservationVerifier(ParkingDatabase.getInstance())));
```
Comment (nicolbidone) [inline app/src/main/java/com/onboarding/parkingsystemjava/activity/ReserveActivity.java:26]: It's necessary to send this as a parameter?

File: app/src/main/java/com/onboarding/parkingsystemjava/entity/Reservation.java [hunk 9]
```diff
@@ -10,7 +10,7 @@
     private String userPassword;
 
     public Reservation() {
-        this.parkingLot = - ConstantUtils.INT_ONE;
+        this.parkingLot = ConstantUtils.INT_MINUS_ONE;
```
Comment (nicolbidone) [inline app/src/main/java/com/onboarding/parkingsystemjava/entity/Reservation.java]: int minus is not a representative name, what is the purpose of the const?

File: app/src/main/java/com/onboarding/parkingsystemjava/utils/ReservationVerifier.java [hunk 10]
```diff
@@ -0,0 +1,70 @@
...
+        Calendar reservationEndDate = reservation.getEndDate();
+        Calendar anotherReservationStartDate = anotherReservation.getStartDate();
+        Calendar anotherReservationEndDate = anotherReservation.getEndDate();
+
+        //start or full overlap
+        if (reservationStartDate.before(anotherReservationStartDate) &&
+                reservationEndDate.after(anotherReservationStartDate)) return true;
+        //end overlap
+        if (reservationStartDate.before(anotherReservationEndDate) &&
+                reservationEndDate.after(anotherReservationEndDate)) return true;
```
Comment (SartiMau) [inline app/src/main/java/com/onboarding/parkingsystemjava/utils/ReservationVerifier.java]: please use culy brackets

File: app/src/main/java/com/onboarding/parkingsystemjava/utils/ReservationVerifier.java (hunk 7 above)
Comment (SartiMau) [reply app/src/main/java/com/onboarding/parkingsystemjava/utils/ReservationVerifier.java]: this still here

**Review by SartiMau — CHANGES_REQUESTED** [review]

File: app/src/main/java/com/onboarding/parkingsystemjava/utils/ReservationVerifier.java (hunk 10 above)
Comment (Faka-GLB) [reply app/src/main/java/com/onboarding/parkingsystemjava/utils/ReservationVerifier.java]: done

File: app/src/main/java/com/onboarding/parkingsystemjava/activity/ReserveActivity.java (hunk 8 above)
Comment (Faka-GLB) [reply app/src/main/java/com/onboarding/parkingsystemjava/activity/ReserveActivity.java:26]: ActivityView requires an Activity in the constructor, and ReserveView extends from ActivityView so i've used **this** sience getActivity() is not aviable

File: app/src/main/java/com/onboarding/parkingsystemjava/entity/Reservation.java (hunk 9 above)
Comment (Faka-GLB) [reply app/src/main/java/com/onboarding/parkingsystemjava/entity/Reservation.java]: In this case is assigned -1 and the verifier checks this value to tell if the user has the parking lot EditText, and the months in Calendar starts in 0  so the const is used to subtract 1 from the month value of the date picker.

File: app/src/main/java/com/onboarding/parkingsystemjava/activity/ReserveActivity.java (hunk 8 above)
Comment (nicolbidone) [reply app/src/main/java/com/onboarding/parkingsystemjava/activity/ReserveActivity.java:26]: hahaha I was asking for the reservation verifier Facu

File: app/src/main/java/com/onboarding/parkingsystemjava/entity/Reservation.java (hunk 9 above)
Comment (nicolbidone) [reply app/src/main/java/com/onboarding/parkingsystemjava/entity/Reservation.java]: It would be nice that the name represent this

File: app/src/main/java/com/onboarding/parkingsystemjava/activity/ReserveActivity.java (hunk 8 above)
Comment (Faka-GLB) [reply app/src/main/java/com/onboarding/parkingsystemjava/activity/ReserveActivity.java:26]: LOL, the verifier is instantiated here for dependency injection reasons, same as the database.

File: app/src/main/java/com/onboarding/parkingsystemjava/entity/Reservation.java (hunk 9 above)
Comment (Faka-GLB) [reply app/src/main/java/com/onboarding/parkingsystemjava/entity/Reservation.java]: done

**Review by BQuinteros — APPROVED** [review]

**Review by nicolbidone — APPROVED** [review]

**Review by SartiMau — APPROVED** [review]

**Review by nicolbidone — APPROVED** [review]

**Review by BQuinteros — APPROVED** [review]

**Review by lucianoeli — APPROVED** [review]

**Review by mauroLonghi — APPROVED** [review]
